File: src/protocol/v1/go/sse_progress.go
package protocolv1

import (
	"bytes"
	"encoding/json"
)

// SSE classes are a fixed vocabulary. Never put provider-controlled event names in diagnostics.
type SSEClass uint8

const (
	SSEStatus SSEClass = iota
	SSEText
	SSEReasoning
	SSETool
	SSEMedia
	SSEItem
	SSETerminal
	SSEEmpty
	SSEOther
	SSEInvalid
	SSEClassCount
)

func (c SSEClass) String() string {
	names := [...]string{"status", "text", "reasoning", "tool", "media", "item", "terminal", "empty", "other", "invalid"}
	if c >= SSEClassCount {
		return "other"
	}
	return names[c]
}

func (c SSEClass) IsOutput() bool { return c >= SSEText && c <= SSEItem }

// ClassifySSEData observes output progress, not response validity or completion proof.
// Scalar payloads become booleans instead of copied strings/trees. Unknown events never
// renew output deadlines. Field matching follows encoding/json's case-insensitive keys.
func ClassifySSEData(data []byte) SSEClass {
	var event progressEvent
	if json.Unmarshal(data, &event) != nil || bytes.Equal(bytes.TrimSpace(data), []byte("null")) {
		return SSEInvalid
	}
	withPayload := func(class SSEClass, present nonemptyText) SSEClass {
		if present {
			return class
		}
		return SSEEmpty
	}
	switch string(event.Type) {
	case "response.completed", "response.failed", "response.incomplete", "error":
		return SSETerminal
	case "response.created", "response.in_progress", "response.queued", "response.metadata",
		"codex.response.metadata", "responsesapi.websocket_timing", "ping", "heartbeat", "keepalive",
		"response.output_item.added", "response.web_search_call.in_progress", "response.web_search_call.searching",
		"response.file_search_call.in_progress", "response.file_search_call.searching",
		"response.code_interpreter_call.in_progress", "response.code_interpreter_call.interpreting",
		"response.image_generation_call.in_progress", "response.image_generation_call.generating":
		return SSEStatus
	case "response.output_text.delta", "response.refusal.delta":
		return withPayload(SSEText, event.Delta)
	case "response.output_text.done":
		return withPayload(SSEText, event.Text)
	case "response.refusal.done":
		return withPayload(SSEText, event.Refusal)
	case "response.reasoning_text.delta", "response.reasoning_summary_text.delta", "response.reasoning.delta":
		return withPayload(SSEReasoning, event.Delta)
	case "response.reasoning_text.done", "response.reasoning_summary_text.done":
		return withPayload(SSEReasoning, event.Text)
	case "response.function_call_arguments.delta", "response.custom_tool_call_input.delta", "response.code_interpreter_call_code.delta":
		return withPayload(SSETool, event.Delta)
	case "response.function_call_arguments.done":
		return withPayload(SSETool, event.Arguments)
	case "response.custom_tool_call_input.done":
		return withPayload(SSETool, event.Input)
	case "response.code_interpreter_call_code.done":
		return withPayload(SSETool, event.Code)
	case "response.audio.delta", "response.output_audio.delta":
		return withPayload(SSEMedia, event.Delta)
	case "response.image_generation_call.partial_image":
		return withPayload(SSEMedia, event.PartialImage)
	case "response.content_part.added", "response.content_part.done":
		return withPayload(SSEText, event.Part.Text || event.Part.Refusal)
	case "response.reasoning_summary_part.added", "response.reasoning_summary_part.done":
		return withPayload(SSEReasoning, event.Part.Text)
	case "response.output_item.done":
		if event.Item.Type == "" || event.Item.unfinished() {
			return SSEEmpty
		}
		switch string(event.Item.Type) {
		case "reasoning":
			return SSEReasoning
		case "function_call", "custom_tool_call", "local_shell_call", "shell_call", "code_interpreter_call", "web_search_call", "file_search_call", "tool_search_call":
			return SSETool
		case "image_generation_call":
			return SSEMedia
		default:
			return SSEItem
		}
	default:
		return SSEOther
	}
}

type progressEvent struct {
	Type         shortText    `json:"type"`
	Delta        nonemptyText `json:"delta"`
	Text         nonemptyText `json:"text"`
	Refusal      nonemptyText `json:"refusal"`
	Arguments    nonemptyText `json:"arguments"`
	Input        nonemptyText `json:"input"`
	Code         nonemptyText `json:"code"`
	PartialImage nonemptyText `json:"partial_image_b64"`
	Item         progressItem `json:"item"`
	Part         progressPart `json:"part"`
}

type nonemptyText bool

func (p *nonemptyText) UnmarshalJSON(data []byte) error {
	*p = len(data) > 2 && data[0] == '"'
	return nil
}

type shortText string

func (s *shortText) UnmarshalJSON(data []byte) error {
	*s = ""
	if len(data) < 2 || len(data) > 1024 || data[0] != '"' {
		return nil
	}
	var value string
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}
	if len(value) <= 128 {
		*s = shortText(value)
	}
	return nil
}

type progressItem struct {
	Type   shortText `json:"type"`
	Status shortText `json:"status"`
}

func (item *progressItem) UnmarshalJSON(data []byte) error {
	*item = progressItem{}
	if len(data) == 0 || data[0] != '{' {
		return nil
	}
	type plain progressItem
	return json.Unmarshal(data, (*plain)(item))
}

func (item progressItem) unfinished() bool {
	switch string(item.Status) {
	case "in_progress", "incomplete", "queued", "searching", "generating", "interpreting":
		return true
	default:
		return false
	}
}

type progressPart struct {
	Text    nonemptyText `json:"text"`
	Refusal nonemptyText `json:"refusal"`
}

func (part *progressPart) UnmarshalJSON(data []byte) error {
	*part = progressPart{}
	if len(data) == 0 || data[0] != '{' {
		return nil
	}
	type plain progressPart
	return json.Unmarshal(data, (*plain)(part))
}

File: src/protocol/v1/rust/src/sse_progress.rs
//! Payload-free event classification shared with Go through the protocol fixtures.
use serde::de::{IgnoredAny, MapAccess, SeqAccess, Visitor};
use serde::{Deserialize, Deserializer};
use std::borrow::Cow;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(usize)]
pub enum Class {
    Status,
    Text,
    Reasoning,
    Tool,
    Media,
    Item,
    Terminal,
    Empty,
    Other,
    Invalid,
}

impl Class {
    pub const COUNT: usize = 10;
    pub const fn as_str(self) -> &'static str {
        [
            "status",
            "text",
            "reasoning",
            "tool",
            "media",
            "item",
            "terminal",
            "empty",
            "other",
            "invalid",
        ][self as usize]
    }
    pub const fn is_output(self) -> bool {
        matches!(
            self,
            Self::Text | Self::Reasoning | Self::Tool | Self::Media | Self::Item
        )
    }
}

pub fn classify(data: &[u8]) -> Class {
    let Ok(event) = serde_json::from_slice::<Fields>(data) else {
        return Class::Invalid;
    };
    let payload = |class, present| if present { class } else { Class::Empty };
    match event.kind.as_str() {
        "response.completed" | "response.failed" | "response.incomplete" => Class::Terminal,
        "error"
        | "response.created"
        | "response.in_progress"
        | "response.queued"
        | "response.metadata"
        | "codex.response.metadata"
        | "responsesapi.websocket_timing"
        | "ping"
        | "heartbeat"
        | "keepalive"
        | "response.output_item.added"
        | "response.web_search_call.in_progress"
        | "response.web_search_call.searching"
        | "response.file_search_call.in_progress"
        | "response.file_search_call.searching"
        | "response.code_interpreter_call.in_progress"
        | "response.code_interpreter_call.interpreting"
        | "response.image_generation_call.in_progress"
        | "response.image_generation_call.generating" => Class::Status,
        "response.output_text.delta" | "response.refusal.delta" => {
            payload(Class::Text, event.delta)
        }
        "response.output_text.done" => payload(Class::Text, event.text),
        "response.refusal.done" => payload(Class::Text, event.refusal),
        "response.reasoning_text.delta"
        | "response.reasoning_summary_text.delta"
        | "response.reasoning.delta" => payload(Class::Reasoning, event.delta),
        "response.reasoning_text.done" | "response.reasoning_summary_text.done" => {
            payload(Class::Reasoning, event.text)
        }
        "response.function_call_arguments.delta"
        | "response.custom_tool_call_input.delta"
        | "response.code_interpreter_call_code.delta" => payload(Class::Tool, event.delta),
        "response.function_call_arguments.done" => payload(Class::Tool, event.arguments),
        "response.custom_tool_call_input.done" => payload(Class::Tool, event.input),
        "response.code_interpreter_call_code.done" => payload(Class::Tool, event.code),
        "response.audio.delta" | "response.output_audio.delta" => {
            payload(Class::Media, event.delta)
        }
        "response.image_generation_call.partial_image" => payload(Class::Media, event.image),
        "response.content_part.added" | "response.content_part.done" => {
            payload(Class::Text, event.part.text || event.part.refusal)
        }
        "response.reasoning_summary_part.added" | "response.reasoning_summary_part.done" => {
            payload(Class::Reasoning, event.part.text)
        }
        "response.output_item.done" => {
            if event.item.kind.is_empty()
                || matches!(
                    event.item.status.as_str(),
                    "in_progress"
                        | "incomplete"
                        | "queued"
                        | "searching"
                        | "generating"
                        | "interpreting"
                )
            {
                return Class::Empty;
            }
            match event.item.kind.as_str() {
                "reasoning" => Class::Reasoning,
                "function_call"
                | "custom_tool_call"
                | "local_shell_call"
                | "shell_call"
                | "code_interpreter_call"
                | "web_search_call"
                | "file_search_call"
                | "tool_search_call" => Class::Tool,
                "image_generation_call" => Class::Media,
                _ => Class::Item,
            }
        }
        _ => Class::Other,
    }
}

#[derive(Default)]
struct Fields {
    kind: String,
    delta: bool,
    text: bool,
    refusal: bool,
    arguments: bool,
    input: bool,
    code: bool,
    image: bool,
    item: Item,
    part: Part,
}

impl<'de> Deserialize<'de> for Fields {
    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
        struct FieldsVisitor;
        impl<'de> Visitor<'de> for FieldsVisitor {
            type Value = Fields;
            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                f.write_str("an SSE event object")
            }
            fn visit_map<M: MapAccess<'de>>(self, mut map: M) -> Result<Fields, M::Error> {
                let mut out = Fields::default();
                while let Some(key) = map.next_key::<Cow<'de, str>>()? {
                    // Match Go's field lookup; duplicate fields consistently use their last value.
                    match key.to_ascii_lowercase().as_str() {
                        "type" => out.kind = map.next_value::<ShortText>()?.0,
                        "delta" => out.delta = map.next_value::<TextPresence>()?.0,
                        "text" => out.text = map.next_value::<TextPresence>()?.0,
                        "refusal" => out.refusal = map.next_value::<TextPresence>()?.0,
                        "arguments" => out.arguments = map.next_value::<TextPresence>()?.0,
                        "input" => out.input = map.next_value::<TextPresence>()?.0,
                        "code" => out.code = map.next_value::<TextPresence>()?.0,
                        "partial_image_b64" => out.image = map.next_value::<TextPresence>()?.0,
                        "item" => out.item = map.next_value()?,
                        "part" => out.part = map.next_value()?,
                        _ => {
                            map.next_value::<IgnoredAny>()?;
                        }
                    }
                }
                Ok(out)
            }
        }
        d.deserialize_map(FieldsVisitor)
    }
}

#[derive(Default)]
struct ShortText(String);
#[derive(Default)]
struct TextPresence(bool);
#[derive(Default)]
struct Item {
    kind: String,
    status: String,
}
#[derive(Default)]
struct Part {
    text: bool,
    refusal: bool,
}

trait Field: Default {
    fn string(_value: &str) -> Self {
        Self::default()
    }
    fn object<'de, M: MapAccess<'de>>(mut map: M) -> Result<Self, M::Error> {
        while map.next_entry::<IgnoredAny, IgnoredAny>()?.is_some() {}
        Ok(Self::default())
    }
}

impl Field for ShortText {
    fn string(value: &str) -> Self {
        Self(if value.len() <= 128 {
            value.to_string()
        } else {
            String::new()
        })
    }
}
impl Field for TextPresence {
    fn string(value: &str) -> Self {
        Self(!value.is_empty())
    }
}
impl Field for Item {
    fn object<'de, M: MapAccess<'de>>(mut map: M) -> Result<Self, M::Error> {
        let mut out = Self::default();
        while let Some(key) = map.next_key::<Cow<'de, str>>()? {
            match key.to_ascii_lowercase().as_str() {
                "type" => out.kind = map.next_value::<ShortText>()?.0,
                "status" => out.status = map.next_value::<ShortText>()?.0,
                _ => {
                    map.next_value::<IgnoredAny>()?;
                }
            }
        }
        Ok(out)
    }
}
impl Field for Part {
    fn object<'de, M: MapAccess<'de>>(mut map: M) -> Result<Self, M::Error> {
        let mut out = Self::default();
        while let Some(key) = map.next_key::<Cow<'de, str>>()? {
            match key.to_ascii_lowercase().as_str() {
                "text" => out.text = map.next_value::<TextPresence>()?.0,
                "refusal" => out.refusal = map.next_value::<TextPresence>()?.0,
                _ => {
                    map.next_value::<IgnoredAny>()?;
                }
            }
        }
        Ok(out)
    }
}

fn field<'de, D: Deserializer<'de>, T: Field>(d: D) -> Result<T, D::Error> {
    struct FieldVisitor<T>(std::marker::PhantomData<T>);
    impl<'de, T: Field> Visitor<'de> for FieldVisitor<T> {
        type Value = T;
        fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            f.write_str("an observed field")
        }
        fn visit_str<E: serde::de::Error>(self, v: &str) -> Result<T, E> {
            Ok(T::string(v))
        }
        fn visit_bool<E: serde::de::Error>(self, _: bool) -> Result<T, E> {
            Ok(T::default())
        }
        fn visit_i64<E: serde::de::Error>(self, _: i64) -> Result<T, E> {
            Ok(T::default())
        }
        fn visit_u64<E: serde::de::Error>(self, _: u64) -> Result<T, E> {
            Ok(T::default())
        }
        fn visit_f64<E: serde::de::Error>(self, _: f64) -> Result<T, E> {
            Ok(T::default())
        }
        fn visit_unit<E: serde::de::Error>(self) -> Result<T, E> {
            Ok(T::default())
        }
        fn visit_map<M: MapAccess<'de>>(self, map: M) -> Result<T, M::Error> {
            T::object(map)
        }
        fn visit_seq<S: SeqAccess<'de>>(self, mut seq: S) -> Result<T, S::Error> {
            while seq.next_element::<IgnoredAny>()?.is_some() {}
            Ok(T::default())
        }
    }
    d.deserialize_any(FieldVisitor(std::marker::PhantomData))
}

macro_rules! observed_field {
    ($($kind:ty),+) => { $(impl<'de> Deserialize<'de> for $kind {
        fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> { field(d) }
    })+ };
}
observed_field!(ShortText, TextPresence, Item, Part);

#[cfg(test)]
#[path = "sse_progress_tests.rs"]
mod tests;
